src/mixer.rs

```diff
@@ -0,0 +1,428 @@
+        try!(acheck!(snd_mixer_selem_register(mixer_handle, ptr::null_mut(), ptr::null_mut())));
+        try!(acheck!(snd_mixer_load(mixer_handle)));
+        Ok(Mixer(mixer_handle))
+    }
+}
+
+/// Closes mixer and frees used resources
+impl Drop for Mixer {
+    fn drop(&mut self) {
+        unsafe { alsa::snd_mixer_close(self.0) };
+    }
+}
+
+/// Wraps [snd_mixer_selem_id_t](http://www.alsa-project.org/alsa-doc/alsa-lib/group___simple_mixer.html)
+#[derive(Copy, Clone)]
+pub struct Elem<'a>{
+    handle: *mut alsa::snd_mixer_elem_t,
+    mixer: &'a Mixer
+}
+
+impl<'a> Elem<'a> {
+    /// Checks if this element is null by checking if the handle is null
+    pub fn is_null(&self) -> bool {
+        self.handle == 0 as *mut alsa::snd_mixer_elem_t
+    }
+}
+
+
+/// Iterator for all elements of mixer
+#[derive(Copy, Clone)]
+pub struct Iter<'a>{
+    last_handle: *mut alsa::snd_mixer_elem_t,
+    mixer: &'a Mixer
+}
+
+impl<'a> Iter<'a> {
+    pub fn new(m: &'a Mixer) -> Iter<'a> {
+        Iter {
+            last_handle: ptr::null_mut(),
+            mixer: m
+        }
+    }
+}
+
+impl<'a> Iterator for Iter<'a> {
+    type Item = Elem<'a>;
+
+    fn next(&mut self) -> Option<Elem<'a>> {
+        let elem = if self.last_handle.is_null() {
+            unsafe { alsa::snd_mixer_first_elem(self.mixer.0) }
+        } else {
+            unsafe { alsa::snd_mixer_elem_next(self.last_handle) }
+        };
+
+        if elem.is_null() {
+            None
+        } else {
+            self.last_handle = elem;
+            Some(Elem { handle: elem, mixer: self.mixer})
+        }
+    }
+
+}
+
+/// Wrapper for `snd_mixer_selem_id_t`, using array of hardcoded length
+// #[derive(Copy, Clone)]
+pub struct SelemId([u8; SELEM_ID_SIZE]);
+
+impl SelemId {
+    /// Creates a new SelemId` of hardcoded size SELEM_ID_SIZE.
+    /// This size is checked against `snd_mixer_selem_id_sizeof`
+    pub fn new(elem: Elem) -> SelemId {
+        // Create empty selem_id and fill from mixer
+        let sid = SelemId::empty();
+        unsafe { alsa::snd_mixer_selem_get_id(elem.handle, sid.as_ptr()) };
+        sid
+    }
+
+    /// Returns an empty (zeroed) SelemId. This id is not a useable id and need to be initialized
+    /// like `SelemId::new()` does
+    pub fn empty() -> SelemId {
+        assert!(unsafe { alsa::snd_mixer_selem_id_sizeof() } as usize <= SELEM_ID_SIZE);
+        // Create empty selem_id and fill from mixer
+        SelemId(unsafe { mem::zeroed() })
+    }
+
+    /// Convert SelemId into ``*mut *mut snd_mixer_selem_id_t` that the alsa call needs.
+    /// See [snd_mixer_selem_id_t](http://www.alsa-project.org/alsa-doc/alsa-lib/group___simple_mixer.html)
+    #[inline]
+    pub fn as_ptr(&self) -> *mut alsa::snd_mixer_selem_id_t {
+        self.0.as_ptr() as *const _ as *mut alsa::snd_mixer_selem_id_t
+    }
+}
+
+/// Wraps an Elem as a Selem
+// #[derive(Copy, Clone)]
+pub struct Selem<'a>(SelemId, Elem<'a>);
```

Comment:
Perhaps I don't understand the difference or similarity between 'Selem' and 'Elem', but 'Elem' wraps 'snd_mixer_elem_t' and 'Selem' wraps 'SelemId' which is of type 'snd_mixer_selem_id_t'. 

Are these the same?